### ISSUE TITLE
2.1.16 PointToVexel replace 1.1 VoxelGenerator，the centerpoint net not work

### ISSUE THREAD

**Issue by gp1234567** [traveller59/spconv#399]
![image](https://user-images.githubusercontent.com/95198402/144163210-292845a1-dd53-4b8a-84de-a7fb4c3e0876.png)
![image](https://user-images.githubusercontent.com/95198402/144163236-2932d383-9b1f-4bc5-a431-086e708a1054.png)
The screenshot above is the 2.1.16 version
![image](https://user-images.githubusercontent.com/95198402/144163405-fc411369-ffa0-4d43-8b9d-44082b4228f6.png)
![image](https://user-images.githubusercontent.com/95198402/144163431-6a3ae740-2ba4-416b-bdcf-5ba7f7e7b857.png)
The screenshot above is the 1.1 version
Is there a problem with my rewriting here？

**Comment by FindDefinition** [comment]
could you provide the detail error log of your problem? your code looks correct and should work.

**Comment by gp1234567** [comment]
Loss decreased slowly than before, and the map was also very low

**Comment by FindDefinition** [comment]
Are you using this [centerpoint](https://github.com/tianweiy/CenterPoint)? I will test difference based on that project.

**Comment by gp1234567** [comment]
yes，but i used OpenPCDet frame

**Comment by FindDefinition** [comment]
I test both voxel generator from centerpoint and spconv 2.x, they produce exactly same result.
what does "the map was also very low" mean?

**Comment by FindDefinition** [comment]
I guess the problem maybe weight init problem, the kaiming initializer only support kcrs layout but default algo of spconv 2.1 use krsc, this makes std of init weight greatly slower. I will fix this in spconv 2.1.17.

**Comment by gp1234567** [comment]
Do you need to rely on pytorch when compiling version 1.0？

**Comment by gp1234567** [comment]
> I guess the problem maybe weight init problem, the kaiming initializer in spconv == 2.1.16 is wrong. I will fix this in spconv 2.1.17. You can use your initializer first.

How to modify it?

**Comment by gp1234567** [comment]
> I test both voxel generator from centerpoint and spconv 2.x, they produce exactly same result. the voxel generator in centerpoint doesn't use spconv 1.x code, they implement it by themselves. Several method to debug this problem:
> 
> 1. Set all algo in conv/pool to ConvAlgo.Native and run again.
> 2. Check inference first, if inference result is correct, then debug train.

Old model train at sp1.0 and inference in sp2.x is good, but in sp2.x, train model problem about loss and eval ap

**Comment by FindDefinition** [comment]
you can change "reset_parameters" to [fixed version](https://github.com/traveller59/spconv/blob/master/spconv/pytorch/conv.py#L168) in your installed spconv for now, or wait for new prebuilt (whole build need ~2h 40min)

**Comment by gp1234567** [comment]
> you can change "reset_parameters" to [fixed version](https://github.com/traveller59/spconv/blob/master/spconv/pytorch/conv.py#L168) in your installed spconv for now, or wait for new prebuilt (whole build need ~2h 40min)

Will you release 2.1.16 prebuilt to fix this problem today？

**Comment by FindDefinition** [comment]
it has been released, but prebuilt need time to build. see [build process](https://github.com/traveller59/spconv/actions/runs/1524302851).

**Comment by gp1234567** [comment]
> it has been released, but prebuilt need time to build. see [build process](https://github.com/traveller59/spconv/actions/runs/1524302851).

![image](https://user-images.githubusercontent.com/95198402/144186397-2e0e5cf5-5fac-4b59-be7b-4c388657cfbd.png)

It's done？？

**Comment by FindDefinition** [comment]
just try 'pip install spconv-cu102==2.1.17 -i https://pypi.python.org/simple' to see if it's done.

**Comment by gp1234567** [comment]
Why is the download speed only 5KB？

**Comment by FindDefinition** [comment]
official python package download from some countries (selected by ```-i https://pypi.python.org/simple'```) (especially China) may be very slow, there are several PyPI mirrors in these countries that can solve this problem but they usually need one day to synchronize newest package from official PyPI.

**Comment by gp1234567** [comment]
I upgrade 2.1.17,but the loss in train process is always not right compare previous versions

**Comment by gp1234567** [comment]
Can i compile sp1.2.1+pytorch1.7+?

**Comment by FindDefinition** [comment]
spconv 1.x is deprecated. You can set all ```algo``` in conv/pool to ```spconv.ConvAlgo.Native``` or try spconv == 2.0.2 to see if problem is resolved. the spconv == 2.0.2 is well tested in waymo dataset.

**Comment by gp1234567** [comment]
![image](https://user-images.githubusercontent.com/95198402/144205272-7971b1a3-f7ed-4153-bfd5-73a06b7caf54.png)

**Comment by FindDefinition** [comment]
spconv 2.0.2 don't support cuda 10.2 ... so you can set algo to ```spconv.ConvAlgo.Native```, this run same algo as spconv 2.0.2

**Comment by gp1234567** [comment]
> spconv 2.0.2 don't support cuda 10.2 ... so you can set algo to `spconv.ConvAlgo.Native`, this run same algo as spconv 2.0.2

Well, I don't understand what you mean. How?

**Comment by FindDefinition** [comment]
there is a [algo argument](https://github.com/traveller59/spconv/blob/v2.1.17/spconv/pytorch/conv.py#L534) in all conv layers and pool layers, default is ```None```, set it to ```spconv.ConvAlgo.Native```.

**Comment by gp1234567** [comment]
Hello, can you add my personal contact information? I haven't solved this problem.(v: gp915342085)

**Comment by FindDefinition** [comment]
I can reproduce your bug in newest OpenPCDet with "PointToVoxel". please wait for bug fix.

**Comment by FindDefinition** [comment]
The bug has been found: I mistakenly thought that "torch.Tensor.numpy" always perform copy, but they actually only return a view of numpy. The VoxelGenerator in spconv 1.x always return a copy of voxels, but PointToVoxel in spconv 2.x return a view of internal storage.
This will be fixed in spconv 2.1.18, PointToVoxel will always return a standalone Tensor instead of a view.

**Comment by gp1234567** [comment]
> The bug has been found: I mistakenly thought that "torch.Tensor.numpy" always perform copy, but they actually only return a view of internal storage. The VoxelGenerator in spconv 1.x always return a copy of voxels, but PointToVoxel in spconv 2.x return a view of internal storage. This will be fixed in spconv 2.1.18, PointToVoxel will always return a standalone Tensor instead of a view. for now you can call .numpy().copy() for every output of PointToVoxel.

But the loss has decreased and is normal, i run centerpoint net in spconv2.0.2_cu111.

**Comment by FindDefinition** [comment]
your bug has been fixed in spconv 2.1.18. spconv 2.0.2 don't have PointToVoxel, so no problem.

**Comment by gp1234567** [comment]
> your bug has been fixed in spconv 2.1.18. spconv 2.0.2 don't have PointToVoxel, so no problem.

Oh, i used Point2VoxelCPU3d replace PointToVoxel.
Thank you for your patience.

**Comment by YoushaaMurhij** [comment]
@FindDefinition , Have you manged to use spconv 2.X voxelizer with CenterPoint?

**Comment by YoushaaMurhij** [comment]
@FindDefinition, @gp1234567 , I tried to replace the implemented voxelizer in CenterPoint by PointToVoxel, which is in Spconv 2.x. but, I am getting an device-side assertion error. Can you please tell me what I am missing here?

Here's my implementation:
```python
from spconv.pytorch.utils import PointToVoxel as VoxelGenerator

self.voxel_generator = VoxelGenerator(
            vsize_xyz=self.voxel_size,
            coors_range_xyz=self.range,
            num_point_features=5,
            max_num_voxels=self.max_voxel_num,
            max_num_points_per_voxel=self.max_points_in_voxel,
            device=self.device
        )

 points = self.points.astype(np.float32)
 pts = torch.from_numpy(points).to(self.device)
 voxels, coords, num_points = self.voxel_generator(pts)
     
 num_voxels = np.array([voxels.shape[0]], dtype=np.int64)
 grid_size = self.voxel_generator.grid_size
 coords = np.pad(coords.cpu().numpy(), ((0, 0), (1, 0)), mode="constant", constant_values=0)

 voxels = voxels.half()
 coords = torch.tensor(coords, dtype=torch.int32, device=self.device)
 num_voxels = torch.tensor(num_voxels, dtype=torch.int32, device=self.device)

 self.inputs = dict(
          voxels=voxels,
          num_points=num_points,
          num_voxels=num_voxels,
          coordinates=coords,
          shape=[grid_size],
      )

  with torch.no_grad():
          outputs = self.net(self.inputs, return_loss=False)[0]

```

Thanks
